### PR TITLE
Intly 2329 3scale

### PR DESCRIPTION
**Pull request by sedroche**

### CODE REVIEW

File: pkg/controller/installation/products/threescale/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+		r.namespace,
+		packageName,
+		"integreatly",
+		[]string{r.namespace},
+		coreosv1alpha1.ApprovalAutomatic)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return v1alpha1.PhaseFailed, err
+	}
+
+	ip, err := r.mpm.GetSubscriptionInstallPlan(packageName, r.namespace)
```
Comment (dimitraz) [inline pkg/controller/installation/products/threescale/reconciler.go]: This error is never checked?

File: pkg/controller/installation/installation_controller.go [hunk 2]
```diff
@@ -220,7 +212,7 @@ func (r *ReconcileInstallation) processStage(instance *v1alpha1.Installation, pr
 		if !(phase == string(v1alpha1.PhaseCompleted)) {
 			incompleteStage = true
 		}
-		reconciler, err := products.NewReconciler(r.context, v1alpha1.ProductName(product), r.client, r.restConfig, r.coreClient, configManager, instance)
```
Comment (philbrookes) [inline pkg/controller/installation/installation_controller.go]: the context has to be the first argument.

File: pkg/controller/installation/products/rhsso/reconciler.go [hunk 3]
```diff
@@ -16,37 +16,66 @@ import (
...
 
-func NewReconciler(coreClient *kubernetes.Clientset, configManager config.ConfigReadWriter, instance *v1alpha1.Installation, mpm marketplace.MarketplaceInterface) (*Reconciler, error) {
+var CustomerAdminUser = &aerogearv1.KeycloakUser{
+	KeycloakApiUser: &aerogearv1.KeycloakApiUser{
+		Enabled:       true,
+		Attributes:    aerogearv1.KeycloakAttributes{},
+		UserName:      "customer-admin",
+		EmailVerified: true,
+		Email:         "customer-admin@example.com",
+		RealmRoles: []string{
```
Comment (philbrookes) [inline pkg/controller/installation/products/rhsso/reconciler.go]: these `realmRoles` don't actually do anything, might as well remove them.

File: pkg/controller/installation/products/rhsso/reconciler.go [hunk 4]
```diff
@@ -16,37 +16,66 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
 	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 var (
-	defaultInstallationNamespace = "rhsso"
-	keycloakName                 = "rhsso"
-	keycloakRealmName            = "openshift"
+	DefaultRhssoNamespace = "rhsso"
```
Comment (philbrookes) [inline pkg/controller/installation/products/rhsso/reconciler.go]: I don't think this should be public, other reconcilers can get this value from the RHSSO config.

File: pkg/controller/installation/products/rhsso/reconciler.go [hunk 5]
```diff
@@ -16,37 +16,66 @@ import (
...
+	customerAdminPassword = "Password1"
+	keycloakName          = "rhsso"
+	KeycloakRealmName     = "openshift"
```
Comment (philbrookes) [inline pkg/controller/installation/products/rhsso/reconciler.go]: let's expose this through the config too?

File: pkg/controller/installation/products/threescale/asserts.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	oauthClient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	coreosv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func assertInstallationSuccessfullyReconciled(installation *v1alpha1.Installation, fakeSigsClient *SigsClientInterfaceMock, fakeThreeScaleClient *ThreeScaleInterfaceMock, fakeOauthClient oauthClient.OauthV1Interface) error {
+	ctx := context.TODO()
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/asserts.go]: we shouldn't be summoning up contexts from nowhere anymore, let's pass in the root context

Comment (philbrookes) [reply pkg/controller/installation/products/threescale/asserts.go]: oh wait, this is to do with unit tests? In that case it's fine!

Comment (philbrookes) [reply pkg/controller/installation/products/threescale/asserts.go]: let's rename the file `_test.go`?

File: pkg/controller/installation/products/threescale/clients.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
+package threescale
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/clients.go]: let's rename the file `_test.go`

File: pkg/controller/installation/products/threescale/clients.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	oauthClient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	coreosv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/testing"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func GetClients(preReqObjects []runtime.Object, scheme *runtime.Scheme, appsv1PreReqs map[string]*appsv1.DeploymentConfig) (*config.Manager, *SigsClientInterfaceMock, appsv1Client.AppsV1Interface, oauthClient.OauthV1Interface, *ThreeScaleInterfaceMock, marketplace.MarketplaceInterface, error) {
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/clients.go]: does this one need to be public?

File: pkg/controller/installation/products/threescale/objects.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
+package threescale
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/objects.go]: this is test related code too right? So let's add _test.go to the filename?

File: pkg/controller/installation/products/threescale/reconciler.go [hunk 10]
```diff
@@ -0,0 +1,592 @@
...
+						AWSRegion: string(bucket.Data["AWS_REGION"]),
+						AWSCredentials: v1.LocalObjectReference{
+							Name: s3SecretName,
+						},
+					},
+				},
+			},
+		},
+	}
+	err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: apim.Name, Namespace: r.namespace}, apim)
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/reconciler.go]: Should we check this before loading in the secret?

File: pkg/controller/installation/products/threescale/reconciler.go [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+			return v1alpha1.PhaseFailed, err
+		}
+	}
+
+	urlSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "credential-rhsso",
+		},
+	}
+	err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: urlSecret.Name, Namespace: r.installation.Spec.NamespacePrefix + rhsso.DefaultRhssoNamespace}, urlSecret)
```
Comment (philbrookes) [inline pkg/controller/installation/products/threescale/reconciler.go]: Let's change this to use `r.ConfigManager.ReadRHSSO().GetNamespace()`

File: pkg/controller/installation/installation_controller.go (hunk 2 above)
Comment (sedroche) [reply pkg/controller/installation/installation_controller.go]: It is passed to `reconcile`, it doesn't seem to be needed here

File: pkg/controller/installation/products/rhsso/reconciler.go (hunk 3 above)
Comment (sedroche) [reply pkg/controller/installation/products/rhsso/reconciler.go]: :+1:

File: pkg/controller/installation/products/rhsso/reconciler.go (hunk 4 above)
Comment (sedroche) [reply pkg/controller/installation/products/rhsso/reconciler.go]: Yeah my bad. I'll update this

File: pkg/controller/installation/products/threescale/clients.go (hunk 8 above)
Comment (sedroche) [reply pkg/controller/installation/products/threescale/clients.go]: probably not. Ill update if it doesn't

File: pkg/controller/installation/installation_controller.go (hunk 2 above)
Comment (philbrookes) [reply pkg/controller/installation/installation_controller.go]: oh, I thought it was moved, didn't realise it was just removed!

File: pkg/controller/installation/products/threescale/objects.go (hunk 9 above)
Comment (sedroche) [reply pkg/controller/installation/products/threescale/objects.go]: Have add `_test.go` to all these

File: pkg/controller/installation/products/threescale/reconciler.go (hunk 10 above)
Comment (sedroche) [reply pkg/controller/installation/products/threescale/reconciler.go]: I want to just reconcile on the objects themselves. if secret is not there, reconcile it. if ApiManager is not there reconcile it. etc.

Comment (philbrookes) [reply pkg/controller/installation/products/threescale/reconciler.go]: if I'm reading the code right, it goes:
1. load in the secret data
2. create the apimanager object in memory
3. see if apimanager exists in api
4. create apimanager if it doesn't exist

I'm suggesting we change the order to:
1. see if apimanager exists, and if not
2.1. load in the secret data
2.2. create the apimanager object in memory
2.3. create apimanager if it doesn't exist

Comment (sedroche) [reply pkg/controller/installation/products/threescale/reconciler.go]: I think we should avoid reconciling objects based on whether another object is there.
We should check if the object exists then reconcile that object.

**Pull request merged**